File: src/main/java/org/gjt/jclasslib/structures/attributes/targettype/TypeArgumentTargetInfo.java
/*
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either
 version 2 of the license, or (at your option) any later version.
 */

package org.gjt.jclasslib.structures.attributes.targettype;

import org.gjt.jclasslib.structures.InvalidByteCodeException;

import java.io.DataInput;
import java.io.DataOutput;
import java.io.IOException;

/**
 * Target info for a <tt>TypeAnnotation</tt> structure with an offset and an argument index.
 */
public class TypeArgumentTargetInfo extends TargetInfo {
    private int offset;
    private int typeArgumentIndex;

    public int getOffset() {
        return offset;
    }

    public void setOffset(int offset) {
        this.offset = offset;
    }

    public int getTypeArgumentIndex() {
        return typeArgumentIndex;
    }

    public void setTypeArgumentIndex(int typeArgumentIndex) {
        this.typeArgumentIndex = typeArgumentIndex;
    }

    @Override
    public void read(DataInput in) throws InvalidByteCodeException, IOException {
        super.read(in);
        offset = in.readUnsignedShort();
        typeArgumentIndex = in.readByte();
    }

    @Override
    public void write(DataOutput out) throws InvalidByteCodeException, IOException {
        super.write(out);
        out.writeShort(offset);
        out.writeByte(typeArgumentIndex);
    }

    @Override
    public int getLength() {
        return 3;
    }

    @Override
    public String getVerbose() {
        return "offset " + offset + ", type argument index " + typeArgumentIndex;
    }
}

File: modules/data/src/main/kotlin/org/gjt/jclasslib/structures/attributes/targettype/TypeArgumentTargetInfo.kt
/*
 This library is free software; you can redistribute it and/or
 modify it under the terms of the GNU General Public
 License as published by the Free Software Foundation; either
 version 2 of the license, or (at your option) any later version.
 */

package org.gjt.jclasslib.structures.attributes.targettype

import java.io.DataInput
import java.io.DataOutput

/**
 * Target info for a TypeAnnotation structure with an offset and an argument index.
 */
class TypeArgumentTargetInfo : TargetInfo() {

    /**
     * The offset.
     */
    var offset: Int = 0

    /**
     * The index of the type argument.
     */
    var typeArgumentIndex: Int = 0

    override fun readData(input: DataInput) {
        offset = input.readUnsignedShort()
        typeArgumentIndex = input.readByte().toInt()
    }

    override fun writeData(output: DataOutput) {
        output.writeShort(offset)
        output.writeByte(typeArgumentIndex)
    }

    override val length: Int
        get() = 3

    override val verbose: String
        get() = "offset $offset, type argument index $typeArgumentIndex"

    override val debugInfo: String
        get() = "with $verbose"
}
